core/operations/management/collection_update.cxx

```diff
@@ -0,0 +1,85 @@
+/* -*- Mode: C++; tab-width: 4; c-basic-offset: 4; indent-tabs-mode: nil -*- */
+/*
+ *   Copyright 2020-2021 Couchbase, Inc.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+#include "collection_update.hxx"
+
+#include "core/utils/json.hxx"
+#include "error_utils.hxx"
+
+#include <regex>
+
+namespace couchbase::core::operations::management
+{
+std::error_code
+collection_update_request::encode_to(encoded_request_type& encoded, http_context& /* context */) const
+{
+    encoded.method = "PATCH";
+    encoded.path = fmt::format("/pools/default/buckets/{}/scopes/{}/collections/{}", bucket_name, scope_name, collection_name);
+    encoded.headers["content-type"] = "application/x-www-form-urlencoded";
+    if (max_expiry.has_value()) {
+        encoded.body.append(fmt::format("maxTTL={}", max_expiry.value()));
+        if (history.has_value()) {
+            encoded.body.append(fmt::format("&history={}", history.value()));
+        }
+        return {};
+    }
+    if (history.has_value()) {
+        encoded.body.append(fmt::format("history={}", history.value()));
+    }
+    return {};
+}
+
+collection_update_response
+collection_update_request::make_response(error_context::http&& ctx, const encoded_response_type& encoded) const
+{
+    collection_update_response response{ std::move(ctx) };
+    if (!response.ctx.ec) {
+        switch (encoded.status_code) {
+            case 400: {
+                if (encoded.body.data().find("Not allowed on this version of cluster") != std::string::npos ||
+                    encoded.body.data().find("Bucket must have storage_mode=magma") != std::string::npos) {
```

Comment:
This is also not ideal, the history should be guarded by a 'nonDedupedHistory' bucket capability check, and return a feature_not_available before sending the request, but I tried adding it with the http_context but it never seemed to update the bucket capabilities, unless I e.g. ran an upsert just before

Reply:
makes sense